### ISSUE TITLE
Fetch only master of index, and in shallow mode

### ISSUE THREAD

**Issue by corneliusweig** [kubernetes-sigs/krew#403]
Currently, we do a plain `git fetch`:

https://github.com/kubernetes-sigs/krew/blob/b303e3ae0179431750deab95be05e7b163286c74/pkg/gitutil/git.go#L52-L54

This has some drawbacks, because it

* fetches all remote branches and tags
* .. and it fetches every commit in each branch.

It would be better, if instead we could

* just fetch the relevant branch (`master`)
* .. and only fetch the tip commit of that branch.

That way, we could keep the `index` folder as lean as possible.

---

This should do the trick:

```go
if err := git(destinationPath, "fetch", "origin", "master", "--verbose", "--depth=1"); err != nil {
		return errors.Wrapf(err, "fetch index at %q failed", destinationPath)
}
```

```go
func EnsureCloned(uri, destinationPath string) error {
	if ok, err := IsGitCloned(destinationPath); err != nil {
		return err
	} else if !ok {
		return exec("", "clone", "--branch=master", "--depth=1", "-v", uri, destinationPath)
	}
	return nil
}
```

**Comment by ahmetb** [comment]
For now, I think this is not a serious issue.

(1) We are still at a very small number of git objects to make any perf issue noticeable. (2) We don't really do branches that would harm clients, and as you said we can later on migrate to a new model without any pain.

**Comment by corneliusweig** [comment]
Yes, the `--depth` is an optimization and not important any time soon.

But only fetching `master` would already be useful. That way we avoid that people pull useless references from origin. For example, there is also a `v0.2.1` branch in my `index` dir.

**Comment by ahmetb** [comment]
Yeah, that's not big of a deal but if you want to fix it, I think we can add `fetch origin master`

**Comment by ahmetb** [comment]
/retitle Fetch only master of index, and in shallow mode

Something else we'd need to consider is going to be private Git repos.

For the centralized index we can hardcode `master` as the branch name to fetch. However, other index repos (even on GitHub) may choose a different main branch name. So this won’t work for them.

I say let's keep things as is, we can refactor over time if needed. I'm not too worried about extra branches getting fetched atm.
/lifecycle frozen

**Comment by slimm609** [comment]
Put up a PR for this issue.   `--single-branch` with out the `--branch=master` only clones the default branch so this supports private repos where the branch is different than master.  Tested with `master` in one repo and `main` in another repo.
